File: naive_bayes/scala/nb.scala
case class DataPoint(features: List[Double], label: Int)

case class Dataset(data: List[DataPoint])

def loadDataset(): Dataset = {
  // Load the dataset
  val data = List[DataPoint]()

  // Your code to load the data from a file or any other source goes here
  // Append each data point as a DataPoint object to the data list

  Dataset(data)
}

def trainNaiveBayes(dataset: Dataset): (List[Double], List[List[Double]]) = {
  val numDataPoints = dataset.data.length
  val numClasses = 3 // Number of classes in your dataset
  val numFeatures = 4 // Number of features in your dataset

  val classCounts = Array.fill(numClasses)(0)
  val priors = Array.fill(numClasses)(0.0)
  val likelihoods = Array.fill(numClasses, numFeatures)(0.0)

  // Count the occurrences of each class label
  dataset.data.foreach { dataPoint =>
    classCounts(dataPoint.label) += 1
  }

  // Calculate priors
  for (i <- 0 until numClasses) {
    priors(i) = classCounts(i) / numDataPoints.toDouble
  }

  // Calculate likelihoods
  for (i <- 0 until numClasses) {
    for (j <- 0 until numFeatures) {
      var featureSum = 0.0
      var featureCount = 0

      // Sum the values of the feature for the current class
      dataset.data.foreach { dataPoint =>
        if (dataPoint.label == i) {
          featureSum += dataPoint.features(j)
          featureCount += 1
        }
      }

      // Calculate the average of the feature for the current class
      likelihoods(i)(j) = featureSum / featureCount.toDouble
    }
  }

  (priors.toList, likelihoods.map(_.toList).toList)
}

def predict(dataPoint: DataPoint, priors: List[Double], likelihoods: List[List[Double]]): Int = {
  val numClasses = priors.length
  val numFeatures = likelihoods(0).length
  var maxPosterior = 0.0
  var predictedLabel = -1

  // Calculate the posterior probability for each class
  for (i <- 0 until numClasses) {
    var posterior = priors(i)

    for (j <- 0 until numFeatures) {
      posterior *= math.exp(-(dataPoint.features(j) - likelihoods(i)(j)).pow(2) / 2)
    }

    // Update the predicted class if the posterior is higher than the current maximum
    if (posterior > maxPosterior) {
      maxPosterior = posterior
      predictedLabel = i
    }
  }

  predictedLabel
}

// Main program
val dataset = loadDataset()
val (priors, likelihoods) = trainNaiveBayes(dataset)

// Example usage: Predict the class label for a new data point
val newDataPoint = DataPoint(List(5.1, 3.5, 1.4, 0.2), 0)

val predictedLabel = predict(newDataPoint, priors, likelihoods)
println("Predicted Label: " + predictedLabel)

File: naive_bayes/javascript/nb.js
class DataPoint {
    constructor(features, label) {
      this.features = features;
      this.label = label;
    }
  }
  
  class Dataset {
    constructor(data) {
      this.data = data;
    }
  }
  
  function loadDataset() {
    // Load the dataset
    const data = [];
  
    // Your code to load the data from a file or any other source goes here
    // Push each data point as a DataPoint object to the data array
  
    return new Dataset(data);
  }
  
  function trainNaiveBayes(dataset) {
    const numDataPoints = dataset.data.length;
    const numClasses = 3; // Number of classes in your dataset
    const numFeatures = 4; // Number of features in your dataset
  
    const classCounts = Array(numClasses).fill(0);
    const priors = Array(numClasses).fill(0);
    const likelihoods = Array(numClasses)
      .fill()
      .map(() => Array(numFeatures).fill(0));
  
    // Count the occurrences of each class label
    dataset.data.forEach((dataPoint) => {
      classCounts[dataPoint.label]++;
    });
  
    // Calculate priors
    for (let i = 0; i < numClasses; i++) {
      priors[i] = classCounts[i] / numDataPoints;
    }
  
    // Calculate likelihoods
    for (let i = 0; i < numClasses; i++) {
      for (let j = 0; j < numFeatures; j++) {
        let featureSum = 0.0;
        let featureCount = 0;
  
        // Sum the values of the feature for the current class
        dataset.data.forEach((dataPoint) => {
          if (dataPoint.label === i) {
            featureSum += dataPoint.features[j];
            featureCount++;
          }
        });
  
        // Calculate the average of the feature for the current class
        likelihoods[i][j] = featureSum / featureCount;
      }
    }
  
    return { priors, likelihoods };
  }
  
  function predict(dataPoint, priors, likelihoods) {
    const numClasses = priors.length;
    const numFeatures = likelihoods[0].length;
    let maxPosterior = 0.0;
    let predictedClass = -1;
  
    // Calculate the posterior probability for each class
    for (let i = 0; i < numClasses; i++) {
      let posterior = priors[i];
  
      for (let j = 0; j < numFeatures; j++) {
        posterior *= Math.exp(
          -Math.pow(dataPoint.features[j] - likelihoods[i][j], 2) / 2
        );
      }
  
      // Update the predicted class if the posterior is higher than the current maximum
      if (posterior > maxPosterior) {
        maxPosterior = posterior;
        predictedClass = i;
      }
    }
  
    return predictedClass;
  }
  
  // Main program
  const dataset = loadDataset();
  const { priors, likelihoods } = trainNaiveBayes(dataset);
  
  // Example usage: Predict the class label for a new data point
  const newDataPoint = new DataPoint([5.1, 3.5, 1.4, 0.2], 0);
  
  const predictedLabel = predict(newDataPoint, priors, likelihoods);
  console.log("Predicted Label:", predictedLabel);
  